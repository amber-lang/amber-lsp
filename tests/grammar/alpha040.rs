use std::fs::read_to_string;

use chumsky::error::Rich;
use insta::assert_debug_snapshot;

use amber_lsp::grammar::{
    alpha040::{lexer::Token, AmberCompiler, GlobalStatement, Spanned},
    LSPAnalysis, ParserResponse,
};

fn tokenize(input: &str) -> Vec<Spanned<Token>> {
    AmberCompiler::new().tokenize(input)
}

fn parse(
    tokens: &[Spanned<Token>],
) -> (Option<Vec<Spanned<GlobalStatement>>>, Vec<Rich<'_, String>>) {
    let ParserResponse {
        ast,
        errors,
        semantic_tokens: _,
    } = AmberCompiler::new().parse(tokens);

    let ast = match ast {
        amber_lsp::grammar::Grammar::Alpha040(ast) => ast,
        _ => panic!("Unexpected AST"),
    };

    (ast, errors)
}

fn parse_unwrap(tokens: &[Spanned<Token>]) -> Vec<Spanned<GlobalStatement>> {
    let (ast, errors) = parse(tokens);
    if !errors.is_empty() {
        panic!("Errors: {:?}", errors);
    }
    ast.unwrap()
}

#[test]
fn test_stdlib_array() {
    let stdlib = read_to_string("resources/alpha040/std/array.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_date() {
    let stdlib = read_to_string("resources/alpha040/std/date.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_env() {
    let stdlib = read_to_string("resources/alpha040/std/env.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_fs() {
    let stdlib = read_to_string("resources/alpha040/std/fs.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_http() {
    let stdlib = read_to_string("resources/alpha040/std/http.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_math() {
    let stdlib = read_to_string("resources/alpha040/std/math.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_text() {
    let stdlib = read_to_string("resources/alpha040/std/text.ab").unwrap();

    let tokens = tokenize(&stdlib);

    assert_debug_snapshot!(tokens);

    assert_debug_snapshot!(parse_unwrap(&tokens));
}

#[test]
fn test_unfinished_function_call() {
    let input = "
    import { array_contains } from \"std/array\"
    
    let x = [1, 2, 3]
    let y = 2
    
    echo array_contains(x, y)
    
    let line = 213
    
    for idx, line in lines(\"text.txt\") {
      echo line
    }
    
    // fun foo(x: Num, y: Text) {
    
    // }
    
    fun foo(x: Num, y: Text) {
    }
    
    // fun abc() {}
    
    array_contains([1, 2, 3],)
    ";

    let tokens = tokenize(input);
    assert_debug_snapshot!(tokens);

    let result = parse(&tokens);

    assert_debug_snapshot!(result);
}
