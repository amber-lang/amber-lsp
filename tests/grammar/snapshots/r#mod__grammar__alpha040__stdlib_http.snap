---
source: tests/grammar/alpha040.rs
expression: parse_unwrap(&tokenize(&stdlib))
---
[
    (
        Import(
            (
                false,
                0..6,
            ),
            (
                "import",
                0..6,
            ),
            (
                ImportAll,
                7..8,
            ),
            (
                "from",
                9..13,
            ),
            (
                "std/env",
                14..23,
            ),
        ),
        0..23,
    ),
    (
        Statement(
            (
                Comment(
                    (
                        DocString(
                            "Downloads a file from a given URL and saves it to a specified path using available command-line tools.",
                        ),
                        25..132,
                    ),
                ),
                25..132,
            ),
        ),
        25..132,
    ),
    (
        Statement(
            (
                Comment(
                    (
                        DocString(
                            "",
                        ),
                        132..136,
                    ),
                ),
                132..136,
            ),
        ),
        132..136,
    ),
    (
        Statement(
            (
                Comment(
                    (
                        DocString(
                            "It checks for the availability of common command-line tools (`curl`, `wget`, and `aria2c`, in order) and uses the first available tool to perform the download.",
                        ),
                        136..300,
                    ),
                ),
                136..300,
            ),
        ),
        136..300,
    ),
    (
        Statement(
            (
                Comment(
                    (
                        DocString(
                            "If none of the tools are available, the function returns `false`.",
                        ),
                        300..370,
                    ),
                ),
                300..370,
            ),
        ),
        300..370,
    ),
    (
        FunctionDefinition(
            [],
            (
                true,
                370..373,
            ),
            (
                "fun",
                374..377,
            ),
            (
                "file_download",
                378..391,
            ),
            [
                (
                    Typed(
                        (
                            false,
                            392..392,
                        ),
                        (
                            "url",
                            392..395,
                        ),
                        (
                            Text,
                            397..401,
                        ),
                    ),
                    392..401,
                ),
                (
                    Typed(
                        (
                            false,
                            403..402,
                        ),
                        (
                            "path",
                            403..407,
                        ),
                        (
                            Text,
                            409..413,
                        ),
                    ),
                    403..413,
                ),
            ],
            Some(
                (
                    Bool,
                    416..420,
                ),
            ),
            [
                (
                    IfChain(
                        (
                            "if",
                            427..429,
                        ),
                        [
                            (
                                IfCondition(
                                    (
                                        IfCondition(
                                            (
                                                FunctionInvocation(
                                                    [],
                                                    (
                                                        "is_command",
                                                        440..450,
                                                    ),
                                                    [
                                                        (
                                                            Text(
                                                                [
                                                                    (
                                                                        Text(
                                                                            (
                                                                                "curl",
                                                                                452..456,
                                                                            ),
                                                                        ),
                                                                        452..456,
                                                                    ),
                                                                ],
                                                            ),
                                                            451..457,
                                                        ),
                                                    ],
                                                    None,
                                                ),
                                                440..458,
                                            ),
                                            (
                                                Block(
                                                    [],
                                                    [
                                                        (
                                                            Expression(
                                                                (
                                                                    Command(
                                                                        [
                                                                            (
                                                                                Trust,
                                                                                473..478,
                                                                            ),
                                                                        ],
                                                                        [
                                                                            (
                                                                                Text(
                                                                                    "$",
                                                                                ),
                                                                                479..480,
                                                                            ),
                                                                            (
                                                                                Text(
                                                                                    "curl",
                                                                                ),
                                                                                481..485,
                                                                            ),
                                                                            (
                                                                                CommandOption(
                                                                                    "-L",
                                                                                ),
                                                                                486..488,
                                                                            ),
                                                                            (
                                                                                CommandOption(
                                                                                    "-o",
                                                                                ),
                                                                                489..491,
                                                                            ),
                                                                            (
                                                                                Text(
                                                                                    "\"",
                                                                                ),
                                                                                492..493,
                                                                            ),
                                                                            (
                                                                                Expression(
                                                                                    (
                                                                                        Var(
                                                                                            (
                                                                                                "path",
                                                                                                494..498,
                                                                                            ),
                                                                                        ),
                                                                                        494..498,
                                                                                    ),
                                                                                ),
                                                                                493..499,
                                                                            ),
                                                                            (
                                                                                Text(
                                                                                    "\"",
                                                                                ),
                                                                                499..500,
                                                                            ),
                                                                            (
                                                                                Text(
                                                                                    "\"",
                                                                                ),
                                                                                501..502,
                                                                            ),
                                                                            (
                                                                                Expression(
                                                                                    (
                                                                                        Var(
                                                                                            (
                                                                                                "url",
                                                                                                503..506,
                                                                                            ),
                                                                                        ),
                                                                                        503..506,
                                                                                    ),
                                                                                ),
                                                                                502..507,
                                                                            ),
                                                                            (
                                                                                Text(
                                                                                    "\"",
                                                                                ),
                                                                                507..508,
                                                                            ),
                                                                            (
                                                                                Text(
                                                                                    "$",
                                                                                ),
                                                                                509..510,
                                                                            ),
                                                                        ],
                                                                        None,
                                                                    ),
                                                                    473..510,
                                                                ),
                                                            ),
                                                            473..510,
                                                        ),
                                                    ],
                                                ),
                                                459..520,
                                            ),
                                        ),
                                        440..520,
                                    ),
                                ),
                                440..520,
                            ),
                            (
                                IfCondition(
                                    (
                                        IfCondition(
                                            (
                                                FunctionInvocation(
                                                    [],
                                                    (
                                                        "is_command",
                                                        529..539,
                                                    ),
                                                    [
                                                        (
                                                            Text(
                                                                [
                                                                    (
                                                                        Text(
                                                                            (
                                                                                "wget",
                                                                                541..545,
                                                                            ),
                                                                        ),
                                                                        541..545,
                                                                    ),
                                                                ],
                                                            ),
                                                            540..546,
                                                        ),
                                                    ],
                                                    None,
                                                ),
                                                529..547,
                                            ),
                                            (
                                                Block(
                                                    [],
                                                    [
                                                        (
                                                            Expression(
                                                                (
                                                                    Command(
                                                                        [
                                                                            (
                                                                                Trust,
                                                                                562..567,
                                                                            ),
                                                                        ],
                                                                        [
                                                                            (
                                                                                Text(
                                                                                    "$",
                                                                                ),
                                                                                568..569,
                                                                            ),
                                                                            (
                                                                                Text(
                                                                                    "wget",
                                                                                ),
                                                                                570..574,
                                                                            ),
                                                                            (
                                                                                Text(
                                                                                    "\"",
                                                                                ),
                                                                                575..576,
                                                                            ),
                                                                            (
                                                                                Expression(
                                                                                    (
                                                                                        Var(
                                                                                            (
                                                                                                "url",
                                                                                                577..580,
                                                                                            ),
                                                                                        ),
                                                                                        577..580,
                                                                                    ),
                                                                                ),
                                                                                576..581,
                                                                            ),
                                                                            (
                                                                                Text(
                                                                                    "\"",
                                                                                ),
                                                                                581..582,
                                                                            ),
                                                                            (
                                                                                CommandOption(
                                                                                    "-P",
                                                                                ),
                                                                                583..585,
                                                                            ),
                                                                            (
                                                                                Text(
                                                                                    "\"",
                                                                                ),
                                                                                586..587,
                                                                            ),
                                                                            (
                                                                                Expression(
                                                                                    (
                                                                                        Var(
                                                                                            (
                                                                                                "path",
                                                                                                588..592,
                                                                                            ),
                                                                                        ),
                                                                                        588..592,
                                                                                    ),
                                                                                ),
                                                                                587..593,
                                                                            ),
                                                                            (
                                                                                Text(
                                                                                    "\"",
                                                                                ),
                                                                                593..594,
                                                                            ),
                                                                            (
                                                                                Text(
                                                                                    "$",
                                                                                ),
                                                                                595..596,
                                                                            ),
                                                                        ],
                                                                        None,
                                                                    ),
                                                                    562..596,
                                                                ),
                                                            ),
                                                            562..596,
                                                        ),
                                                    ],
                                                ),
                                                548..606,
                                            ),
                                        ),
                                        529..606,
                                    ),
                                ),
                                529..606,
                            ),
                            (
                                IfCondition(
                                    (
                                        IfCondition(
                                            (
                                                FunctionInvocation(
                                                    [],
                                                    (
                                                        "is_command",
                                                        615..625,
                                                    ),
                                                    [
                                                        (
                                                            Text(
                                                                [
                                                                    (
                                                                        Text(
                                                                            (
                                                                                "aria2c",
                                                                                627..633,
                                                                            ),
                                                                        ),
                                                                        627..633,
                                                                    ),
                                                                ],
                                                            ),
                                                            626..634,
                                                        ),
                                                    ],
                                                    None,
                                                ),
                                                615..635,
                                            ),
                                            (
                                                Block(
                                                    [],
                                                    [
                                                        (
                                                            Expression(
                                                                (
                                                                    Command(
                                                                        [
                                                                            (
                                                                                Trust,
                                                                                650..655,
                                                                            ),
                                                                        ],
                                                                        [
                                                                            (
                                                                                Text(
                                                                                    "$",
                                                                                ),
                                                                                656..657,
                                                                            ),
                                                                            (
                                                                                Text(
                                                                                    "aria2c",
                                                                                ),
                                                                                658..664,
                                                                            ),
                                                                            (
                                                                                Text(
                                                                                    "\"",
                                                                                ),
                                                                                665..666,
                                                                            ),
                                                                            (
                                                                                Expression(
                                                                                    (
                                                                                        Var(
                                                                                            (
                                                                                                "url",
                                                                                                667..670,
                                                                                            ),
                                                                                        ),
                                                                                        667..670,
                                                                                    ),
                                                                                ),
                                                                                666..671,
                                                                            ),
                                                                            (
                                                                                Text(
                                                                                    "\"",
                                                                                ),
                                                                                671..672,
                                                                            ),
                                                                            (
                                                                                CommandOption(
                                                                                    "-d",
                                                                                ),
                                                                                673..675,
                                                                            ),
                                                                            (
                                                                                Text(
                                                                                    "\"",
                                                                                ),
                                                                                676..677,
                                                                            ),
                                                                            (
                                                                                Expression(
                                                                                    (
                                                                                        Var(
                                                                                            (
                                                                                                "path",
                                                                                                678..682,
                                                                                            ),
                                                                                        ),
                                                                                        678..682,
                                                                                    ),
                                                                                ),
                                                                                677..683,
                                                                            ),
                                                                            (
                                                                                Text(
                                                                                    "\"",
                                                                                ),
                                                                                683..684,
                                                                            ),
                                                                            (
                                                                                Text(
                                                                                    "$",
                                                                                ),
                                                                                685..686,
                                                                            ),
                                                                        ],
                                                                        None,
                                                                    ),
                                                                    650..686,
                                                                ),
                                                            ),
                                                            650..686,
                                                        ),
                                                    ],
                                                ),
                                                636..696,
                                            ),
                                        ),
                                        615..696,
                                    ),
                                ),
                                615..696,
                            ),
                            (
                                Else(
                                    (
                                        Else(
                                            (
                                                "else",
                                                705..709,
                                            ),
                                            (
                                                Block(
                                                    [],
                                                    [
                                                        (
                                                            Return(
                                                                (
                                                                    "return",
                                                                    724..730,
                                                                ),
                                                                Some(
                                                                    (
                                                                        Boolean(
                                                                            (
                                                                                false,
                                                                                731..736,
                                                                            ),
                                                                        ),
                                                                        731..736,
                                                                    ),
                                                                ),
                                                            ),
                                                            724..736,
                                                        ),
                                                    ],
                                                ),
                                                710..746,
                                            ),
                                        ),
                                        705..746,
                                    ),
                                ),
                                705..746,
                            ),
                        ],
                    ),
                    427..752,
                ),
                (
                    Return(
                        (
                            "return",
                            758..764,
                        ),
                        Some(
                            (
                                Boolean(
                                    (
                                        true,
                                        765..769,
                                    ),
                                ),
                                765..769,
                            ),
                        ),
                    ),
                    758..769,
                ),
            ],
        ),
        370..771,
    ),
]
